### PR TITLE
[fix] Attach train+val dataloaders to trainer in trainer loop

### LINKED ISSUE #7208: Using reload_dataloaders_every_epoch=True and num_sanity_val_steps=0 can lead to the validation loop being skipped
## 🐛 Bug

Inside the training loop, we incorrectly skip running evaluation when `reload_dataloaders_every_epoch=True` and `num_sanity_val_steps=0`. With these settings, we defer setting the validation dataloader on the trainer until the evaluation loop is run from inside the training loop. However, this is too late as the training loop depends on the validation dataloader settings being set in order to even determine *whether* we run the evaluation loop at all.

This means it's possible to have these states set inside of the training loop when determining whether to run the evaluation loop:
```
is_last_batch=True
should_check_val=True
num_val_batches=[]
should_skip_eval=True
disable_validation=False
should_train_only=True
```
`should_skip_eval=True` when `self.trainer.num_val_batches` isn't set: In this instance `trainer.num_val_batches=[]` .
https://github.com/PyTorchLightning/pytorch-lightning/blob/44d775fccfb825561937f6fa03fe258af25c2b83/pytorch_lightning/trainer/training_loop.py#L551

This points out that `should_check_val` and `should_train_only` were not consistent with each other :( 

https://github.com/PyTorchLightning/pytorch-lightning/pull/6075 changed the order with which we call `run_evaluation` inside the training loop. Before, this was covered up by luck because of the ordering. After the swap occurred there, this has been broken. 

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1z9ln3gYBK-VGidNPdUE2UgE0ISAgjLpu?usp=sharing

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Expected behavior

Checkpointing should still work as expected because we run the evaluation loop when expected

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7208 . 

The issue occurs because the validation dataloader is never attached to the trainer by the time we reach this stage in the training loop: https://github.com/PyTorchLightning/pytorch-lightning/blob/44d775fccfb825561937f6fa03fe258af25c2b83/pytorch_lightning/trainer/training_loop.py#L558-L560

Currently this is conditional based on this logic, which is called once at the start of training: https://github.com/PyTorchLightning/pytorch-lightning/blob/44d775fccfb825561937f6fa03fe258af25c2b83/pytorch_lightning/trainer/trainer.py#L604

Removing the `reload_dataloaders_every_epoch` check here gives me consistent behavior again.

I don't think this is a fully proper fix - we should be attaching these dataloaders to the trainer more clearly, and operating them on them for reloads later on in the training loop. 

This is a behavior change that resulted after #6075. Before, `run_evaluation` ran, which populated the dataloader settings on the trainer, and then the checkpoint callback ran. After, the checkpoint callback ran first, before run_evaluation. the check for whether we run evaluation depends on the val dataloader being present, which was being set inside of run_evaluation.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -1227,21 +1227,22 @@ def validation_step(self, batch, batch_idx):
...
-        "train_dataloader",
-        "val_dataloader",
-        "train_dataloader",
-        "val_dataloader",
-        "train_dataloader",
-        "val_dataloader",
-        "test_dataloader",
+        'train_dataloader',
+        'val_dataloader',
+        'val_dataloader',
```
Comment (ananthsub) [inline tests/trainer/test_dataloaders.py:1240]: this is an extra reload coming from the training loop calling run_evaluation. we could special case this by avoiding the reload if epoch = 0 but that feels hacky too: https://github.com/PyTorchLightning/pytorch-lightning/blob/68eac4d9486b95f8d1b09ae5468885c48d3d5cd0/pytorch_lightning/trainer/evaluation_loop.py#L63-L64

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -200,10 +200,10 @@ def on_train_batch_end(self, epoch_output, batch_end_outputs, batch, batch_idx,
         self.trainer.logger_connector.on_train_batch_end()
 
     def reset_train_val_dataloaders(self, model):
-        if self.trainer.train_dataloader is None or not self.trainer.reload_dataloaders_every_epoch:
+        if self.trainer.train_dataloader is None:
             self.trainer.reset_train_dataloader(model)
 
-        if self.trainer.val_dataloaders is None and not self.trainer.reload_dataloaders_every_epoch:
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:206]: Should we remove the val reload and leave its responsibility to the val_loop instead ?

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -200,10 +200,10 @@ def on_train_batch_end(self, epoch_output, batch_end_outputs, batch, batch_idx,
         self.trainer.logger_connector.on_train_batch_end()
 
     def reset_train_val_dataloaders(self, model):
-        if self.trainer.train_dataloader is None or not self.trainer.reload_dataloaders_every_epoch:
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:203]: Can you add a test to make sure different dataloaders get properly reloaded when reload_dataloaders_every_epoch=True ?

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1240]: We are doing this for train already: https://github.com/PyTorchLightning/pytorch-lightning/blob/44d775fccfb825561937f6fa03fe258af25c2b83/pytorch_lightning/trainer/training_loop.py#L169

**Comment by edenlightning** [comment]
@justusschock @awaelchli

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:206]: The val dataloader must be set before training because the train loop depends on it for determining whether to run evaluation

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (ananthsub) [reply tests/trainer/test_dataloaders.py:1240]: @tchaton does checking `trainer.current_epoch` work if someone is running multiple validation epochs per training epoch? is there a risk that we don't reload the dataloaders on each new validation pass?

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:203]: could you describe what's missing from the tests here? https://github.com/PyTorchLightning/pytorch-lightning/blob/ac7d6a35c3d36267b1f0f66ee6609ac40d244dba/tests/trainer/test_dataloaders.py#L1201

**Comment by ananthsub** [comment]
@awaelchli @justusschock @tchaton what do you think?

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (justusschock) [reply tests/trainer/test_dataloaders.py:1240]: Yes, I think there is a risk on that. We could add a separate counter for validation epochs though

**Review by justusschock — COMMENTED** [review]
Looks good to me.

**Review by justusschock — APPROVED** [review]

**Comment by ananthsub** [comment]
@tchaton @awaelchli @Borda mind taking a look?

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (Borda) [reply tests/trainer/test_dataloaders.py:1239]: can you add a comment to the code with this explanation?

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -1209,21 +1219,32 @@ def test_dataloaders_load_every_epoch_no_sanity_check(tmpdir):
...
         'train_dataloader',
         'val_dataloader',
+        # This has subsequent calls to val_dataloader
+        # because the training loop runs the evaluation loop,
+        # which reloads the val dataloader again.
+        # We cannot yet rely on trainer.current_epoch=0 to skip reloading
+        # the val dataloader on the first epoch because this only tracks the training epoch
+        # meaning multiple passes through the validation data within a single training epoch
+        # would not have the datalodaer reloaded.
+        # This breaks the assumption behind reload_dataloaders_every_epoch=True
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1246]:
```suggestion
        # would not have the dataloader reloaded.
        # This breaks the assumption behind reload_dataloaders_every_epoch=True
```

**Review by awaelchli — APPROVED** [review]

**Comment by ananthsub** [comment]
@carmocca @Borda @tchaton mind taking a look?

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
